src/box/lua/net_box.lua

```diff
@@ -238,20 +237,21 @@ local stream_spaces_mt = {
     __serialize = stream_spaces_serialize
 }
 
--- This callback is invoked in a new fiber upon receiving a SHUTDOWN packet
+-- This callback is invoked in a new fiber upon receiving 'box.shutdown' event
 -- from a remote host. It runs on_shutdown triggers and then gracefully
 -- terminates the connection.
 local function graceful_shutdown(remote)
+    remote._shutdown_pending = true
     local ok, err = pcall(remote._on_shutdown.run, remote._on_shutdown, remote)
     if not ok then
         log.error(err)
     end
     -- While the triggers were running, the connection could have been closed
     -- and even reestablished (if reconnect_after is set), in which case we
     -- must not initiate a graceful shutdown.
```

Comment:
What if it was re-established, and then second shutdown was started? Then the first one will run `graceful_shutdown()` and will mess with the new connection. Probably this should be a generation counter instead of a flag. If this is really a bug. But if it is not, you might want to consider calling it `_is_shutdown_pending` (`is` suffix for flags is commonly used in our code), I am fine with either name.

Reply:
The second shutdown callback won't start until the first one returns - this is how the watcher infrastructure works. I have a test case checking this, see `graceful_shutdown.test_shutdown_aborted_on_reconnect_2`.